### ISSUE TITLE
Installation issue: curl

### ISSUE THREAD

**Issue by mathomp4** [spack/spack#9958]
### Steps to reproduce the issue

First, I am actually installing `curl%gcc@8.2.0` but I've defaulted to `gcc@8.2.0` in `packages.yaml`. My code is (eventually) Fortran so `clang` can be an adventure.

```console
$ spack install curl
==> Installing curl
==> Searching for binary cache of curl
==> Warning: No Spack mirrors are currently configured
==> No binary for curl found: installing from source
==> Fetching http://curl.haxx.se/download/curl-7.60.0.tar.bz2
######################################################################## 100.0%
==> Staging archive: /Users/mathomp4/spack/var/spack/stage/curl-7.60.0-tngzkrdmhpblo4o6cscyl2idlwmjolnf/curl-7.60.0.tar.bz2
==> Created stage in /Users/mathomp4/spack/var/spack/stage/curl-7.60.0-tngzkrdmhpblo4o6cscyl2idlwmjolnf
==> No patches needed for curl
==> Building curl [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j4'

9 errors found in build log:
     909      CC       vtls/libcurl_la-darwinssl.lo
     910      CC       vtls/libcurl_la-gskit.lo
     911      CC       vtls/libcurl_la-mbedtls.lo
     912    In file included from /System/Library/Frameworks/Security.framework/Headers/AuthSession.h:32,
     913                     from /System/Library/Frameworks/Security.framework/Headers/Security.h:43,
     914                     from vtls/darwinssl.c:44:
  >> 915    /System/Library/Frameworks/Security.framework/Headers/Authorization.h:193:7: error: variably modified 'bytes' at file scope
     916      char bytes[kAuthorizationExternalFormLength];
     917           ^~~~~
     918    vtls/darwinssl.c: In function 'darwinssl_version_from_curl':
  >> 919    vtls/darwinssl.c:1258:10: error: implicit declaration of function '__builtin_available'; did you mean '__builtin_scalbl'? [-Werror=implicit-
            function-declaration]
     920           if(__builtin_available(macOS 10.13, iOS 11.0, *)) {
     921              ^~~~~~~~~~~~~~~~~~~
     922              __builtin_scalbl
  >> 923    vtls/darwinssl.c:1258:30: error: 'macOS' undeclared (first use in this function)
     924           if(__builtin_available(macOS 10.13, iOS 11.0, *)) {
     925                                  ^~~~~
     926    vtls/darwinssl.c:1258:30: note: each undeclared identifier is reported only once for each function it appears in
  >> 927    vtls/darwinssl.c:1258:35: error: expected ')' before numeric constant
     928           if(__builtin_available(macOS 10.13, iOS 11.0, *)) {
     929                                       ^~~~~~
     930                                       )
     931    vtls/darwinssl.c: In function 'set_ssl_version_min_max':
  >> 932    vtls/darwinssl.c:1282:26: error: 'macOS' undeclared (first use in this function)
     933       if(__builtin_available(macOS 10.13, iOS 11.0, *)) {
     934                              ^~~~~
  >> 935    vtls/darwinssl.c:1282:31: error: expected ')' before numeric constant
     936       if(__builtin_available(macOS 10.13, iOS 11.0, *)) {
     937                                   ^~~~~~
     938                                   )
     939    vtls/darwinssl.c: In function 'darwinssl_connect_step1':
  >> 940    vtls/darwinssl.c:1434:30: error: 'macOS' undeclared (first use in this function)
     941           if(__builtin_available(macOS 10.13, iOS 11.0, *)) {
     942                                  ^~~~~
  >> 943    vtls/darwinssl.c:1434:35: error: expected ')' before numeric constant
     944           if(__builtin_available(macOS 10.13, iOS 11.0, *)) {
     945                                       ^~~~~~
     946                                       )
     947    vtls/darwinssl.c: In function 'show_verbose_server_cert':
  >> 948    vtls/darwinssl.c:2517:6: error: 'SecTrustEvaluateAsync' undeclared (first use in this function); did you mean 'SecTrustEvaluate'?
     949       if(SecTrustEvaluateAsync != NULL) {
     950          ^~~~~~~~~~~~~~~~~~~~~
     951          SecTrustEvaluate
     952    cc1: some warnings being treated as errors
     953    make[2]: *** [Makefile:2175: vtls/libcurl_la-darwinssl.lo] Error 1
     954    make[2]: *** Waiting for unfinished jobs....

See build log for details:
  /Users/mathomp4/spack/var/spack/stage/curl-7.60.0-tngzkrdmhpblo4o6cscyl2idlwmjolnf/curl-7.60.0/spack-build.out
```

### Platform and user environment

Please report your OS here:
```commandline
$ uname -a
Darwin gs6101-parcel.ndc.nasa.gov 17.7.0 Darwin Kernel Version 17.7.0: Wed Oct 10 23:06:14 PDT 2018; root:xnu-4570.71.13~1/RELEASE_X86_64 x86_64
$ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.13.6
BuildVersion:	17G3025
``` 
And my `packages.yaml` tries to use the system ssl, though I'm still a bit confused on how to do it on macOS as it's Libre SSL and not Open SSL. Someone on slack said to just pick a version of Open SSL and have spack pretend the system is that version.
```commandline
[mathomp4:~/.spack] 7d20h32m39s $ cat packages.yaml
packages:
   all:
     compiler: [gcc@8.2.0]
     providers:
       mpi: [openmpi,mpich]
   openssl:
     paths:
       openssl@1.0.2o: /usr
     buildable: False
   hdf5:
     variants: +fortran+szip
   netcdf:
     variants: +hdf4+dap
```
I've tried building with and without the `openssl` lines in `packages.yaml` and both fail the same way.

And here is my `compilers.yaml`:

```commandline
[mathomp4:~/.spack] 7d20h33m12s $ cat darwin/compilers.yaml
compilers:
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: highsierra
    paths:
      cc: /usr/bin/clang
      cxx: /usr/bin/clang++
      f77: /usr/local/bin/gfortran
      fc: /usr/local/bin/gfortran
    spec: clang@10.0.0-apple
    target: x86_64
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: highsierra
    paths:
      cc: null
      cxx: null
      f77: /usr/local/bin/gfortran
      fc: /usr/local/bin/gfortran
    spec: gcc@6.3.0
    target: x86_64
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: [gcc-gfortran/8.2.0]
    operating_system: highsierra
    paths:
      cc: /Users/mathomp4/installed/Core/gcc-gfortran/8.2.0/bin/gcc
      cxx: /Users/mathomp4/installed/Core/gcc-gfortran/8.2.0/bin/g++
      f77: /Users/mathomp4/installed/Core/gcc-gfortran/8.2.0/bin/gfortran
      fc: /Users/mathomp4/installed/Core/gcc-gfortran/8.2.0/bin/gfortran
    spec: gcc@8.2.0
    target: x86_64
```

### Additional information

Per issue template, some additional information:
```console
[mathomp4:~/.spack] 6d21h1m10s 1 $ spack spec --install-status curl
Input spec
--------------------------------
 -   curl

Concretized
--------------------------------
 -   curl@7.60.0%gcc@8.2.0+darwinssl~libssh~libssh2~nghttp2 arch=darwin-highsierra-x86_64
[+]      ^zlib@1.2.11%gcc@8.2.0+optimize+pic+shared arch=darwin-highsierra-x86_64
```

I have also rebuilt the package with the following options:
```console
$ spack -d install -j 1 curl
```
and will attach the files next.

**Comment by mathomp4** [comment]
[spack-cc-curl-tngzkrd.in.log](https://github.com/spack/spack/files/2620448/spack-cc-curl-tngzkrd.in.log)
[spack-cc-curl-tngzkrd.out.log](https://github.com/spack/spack/files/2620449/spack-cc-curl-tngzkrd.out.log)

**Comment by mathomp4** [comment]
I managed to hit this all again in my latest attempts when it was theorized on slack that another issue I had was trying to use the built-in `curl` and the `-L/usr/lib` it then added nuked some things.

So, based on suggestion by @pramodk, I tried building `curl` and, yup darwinssl issues. I then tried to install `curl~darwinssl` and it worked, but I had to allow spack to install `openssl`. I couldn't figure out how to get it to use the builtin `LibreSSL`.

**Comment by sethrj** [comment]
@mathomp4 I've hit the same set of errors [when building glib using gcc](https://github.com/spack/spack/pull/10944#discussion_r270607962): GCC8 fails to compile some of the cocoa frameworks even though it can preprocess the header files. I highly recommend a mixed build system using the apple `clang` to build C/C++/Obj-C libraries and GCC8 only to provide the fortran compilers. I've been doing that for years and find it to be quite stable, and my attempts to get a fully-GCC8 toolchain on Spack on macOS 10.13 have failed.

**Comment by mathomp4** [comment]
@sethrj Glad I'm not alone. I thought at one point I figured out a workaround (use bash instead of tcsh), but I think in the end even that didn't help.

Perhaps things are better now for `clang`. In the past, I had issues building ESMF with `clang` (or, well, `clang++`) but I haven't tried it in a while.

**Comment by climbfuji** [comment]
@mathomp4 Is this still an issue?

**Comment by mathomp4** [comment]
Probably not. A bit ancient. I'll close and reopen if needed.
